### DISCUSSION TITLE
How about building your own dht bootstrap node?

### DISCUSSION THREAD

**Discussion by hxx258456** [libp2p/go-libp2p#2926] [Q&A]
```go
func CreateNode(privateKey crypto.PrivKey, port int) (host.Host, error) {

	options := []libp2p.Option{
		libp2p.EnableAutoNATv2(),
		libp2p.NATPortMap(),
		libp2p.EnableNATService(),
		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
			//return dht.New(context.Background(), h, dht.BootstrapPeers(dht.GetDefaultBootstrapPeerAddrInfos()...))
			return dht.New(context.Background(), h, dht.Mode(dht.ModeServer))
		}),
		libp2p.ForceReachabilityPublic(),
		libp2p.FallbackDefaults,
	}

	if privateKey != nil {
		options = append(options, libp2p.Identity(privateKey))
	}
	return libp2p.New(options...)
}
```
I'm trying to build my own dht boostrap node, but when I use the libp2p configuration above I find that other nodes can't connect to it.

**Comment by Stebalien** [comment]
How did you configure your other nodes? Does your bootstrap node have a static IP address/port?

I ask because I see a bunch of NAT options and putting a bootstrap node behind a NAT generally doesn't work well.

**Reply by hxx258456** [reply]
The other nodes I created using the same function, except that after creating them I added a step to connect to that node, but an error occurs while connecting
I took your advice and changed my node configuration.
```go
func CreateNode(privateKey crypto.PrivKey, port int) (host.Host, error) {

	options := []libp2p.Option{
		libp2p.ListenAddrStrings("/ip4/0.0.0.0/tcp/9200"),
		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
			//return dht.New(context.Background(), h, dht.BootstrapPeers(dht.GetDefaultBootstrapPeerAddrInfos()...))
			return dht.New(context.Background(), h, dht.Mode(dht.ModeServer))
		}),
	}

	if privateKey != nil {
		options = append(options, libp2p.Identity(privateKey))
	}
	return libp2p.New(options...)
}
```
I printed the node's addrs, and I noticed that there is no public address in them.
They look like this.
```shell
[/ip4/127.0.0.1/tcp/9200 /ip4/172.20.157.60/tcp/9200]
```

**Reply by hxx258456** [reply]
I forgot to mention that this bootstrap node is running on a cloud service and the firewall port 9200 is open.

**Reply by hxx258456** [reply]
I think the problem is this, there aren't enough nodes for him to do nat traversal, causing him to not detect his public address.

**Reply by hxx258456** [reply]
Must a node's public address depend on other nodes for detection? Can't you specify it yourself?

**Reply by Stebalien** [reply]
You can pass a custom `AddressFactory` option to the libp2p constructor. This function:

1. Takes in the addresses the libp2p node thinks it has.
2. Returns the addresses you want to announce.

**Reply by Stebalien** [reply]
But note: if you're implementing a custom bootstrap node, I'm not sure how that'll help you. The peers connecting to your bootstrap node will need to know the public IP address of your bootstrap node.

> but an error occurs while connecting

Do they successfully connect to the bootstrap node? Or do they have difficulty connecting to each-other afterwards.

**Comment by cpeliciari** [comment]
When starting a bootstrap, don’t forget to initiate it.

[specs](https://github.com/libp2p/specs/tree/master/kad-dht#bootstrap-process)

```go
package main

import (
	"context"
	"github.com/libp2p/go-libp2p"
	dht "github.com/libp2p/go-libp2p-kad-dht"
	"github.com/libp2p/go-libp2p/core/crypto"
	"github.com/libp2p/go-libp2p/core/host"
	"github.com/libp2p/go-libp2p/core/routing"
)

type Config struct {
	PrivateKey crypto.PrivKey
	Bootstrap bool
}

func CreateNode(ctx context.Context, cfg *Config) (h host.Host, ddht *dht.IpfsDHT, err error) {

	options := []libp2p.Option{
		libp2p.EnableAutoNATv2(),
		libp2p.NATPortMap(),
		libp2p.EnableNATService(),
		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
			ddht, err = dht.New(ctx, h)
			if err != nil {
				return nil, err
			}

			return ddht, nil
		}),
		libp2p.ForceReachabilityPublic(),
		libp2p.FallbackDefaults,
	}

	if cfg.PrivateKey != nil {
		options = append(options, libp2p.Identity(cfg.PrivateKey))
	}

	h, err = libp2p.New(options...)
	if err != nil {
		return nil, ddht,err
	}

	if cfg.Bootstrap {
		if err := ddht.Bootstrap(ctx); err != nil {
			return nil,ddht, err
		}
	}

	return h, ddht,err
}
```
